File: app/src/pitstop/java/com/pitstop/database/LocalPidStorage.java
package com.pitstop.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.pitstop.models.Pid;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Paul Soladoye on 4/1/2016.
 */
public class LocalPidStorage {

    private final String TAG = getClass().getSimpleName();

    // PID_DATA table create statement
    public static final String CREATE_TABLE_PID_DATA = "CREATE TABLE IF NOT EXISTS "
            + TABLES.PID.TABLE_NAME + "(" + TABLES.COMMON.KEY_ID + " INTEGER PRIMARY KEY,"
            + TABLES.PID.KEY_DATANUM + " TEXT,"
            + TABLES.PID.KEY_TIMESTAMP + " TEXT,"
            + TABLES.PID.KEY_RTCTIME + " TEXT,"
            + TABLES.PID.KEY_PIDS + " TEXT,"
            + TABLES.PID.KEY_TRIP_ID + " INTEGER,"
            + TABLES.PID.KEY_TRIP_ID_RAW +" INTEGER,"
            + TABLES.PID.KEY_DEVICE_ID +" TEXT,"
            + TABLES.PID.KEY_MILEAGE + " REAL,"
            + TABLES.PID.KEY_CALCULATED_MILEAGE + " REAL,"
            + TABLES.COMMON.KEY_CREATED_AT + " DATETIME" + ")";

    private LocalDatabaseHelper databaseHelper;


    public LocalPidStorage(LocalDatabaseHelper databaseHelper) {
        this.databaseHelper = databaseHelper;
    }

    /**
     * Create pid data
     */
    public void createPIDData(Pid pidData) {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();

        ContentValues values = new ContentValues();
        values.put(TABLES.PID.KEY_DATANUM, pidData.getDataNumber());
        values.put(TABLES.PID.KEY_RTCTIME, pidData.getRtcTime());
        values.put(TABLES.PID.KEY_TIMESTAMP, pidData.getTimeStamp());
        values.put(TABLES.PID.KEY_TRIP_ID_RAW, pidData.getTripIdRaw());
        values.put(TABLES.PID.KEY_TRIP_ID, pidData.getTripId());
        values.put(TABLES.PID.KEY_PIDS, pidData.getPids());
        values.put(TABLES.PID.KEY_MILEAGE, pidData.getMileage());
        values.put(TABLES.PID.KEY_CALCULATED_MILEAGE, pidData.getCalculatedMileage());
        values.put(TABLES.PID.KEY_DEVICE_ID, pidData.getDeviceId());

        db.insert(TABLES.PID.TABLE_NAME, null, values);

    }

    /**
     * Get all pid data
     */
    public List<Pid> getAllPidDataEntries() {
        List<Pid> pidDataEntries = new ArrayList<>();

        String selectQuery = "SELECT * FROM " + TABLES.PID.TABLE_NAME;

        SQLiteDatabase db = databaseHelper.getReadableDatabase();
        Cursor c = db.rawQuery(selectQuery,null);

        if(c.moveToFirst()) {
            do {
                Pid pidData = new Pid();
                pidData.setId(c.getInt(c.getColumnIndex(TABLES.COMMON.KEY_ID)));
                pidData.setDataNumber(c.getString(c.getColumnIndex(TABLES.PID.KEY_DATANUM)));
                pidData.setRtcTime(c.getString(c.getColumnIndex(TABLES.PID.KEY_RTCTIME)));
                pidData.setTimeStamp(c.getString(c.getColumnIndex(TABLES.PID.KEY_TIMESTAMP)));
                pidData.setTripIdRaw(c.getLong(c.getColumnIndex(TABLES.PID.KEY_TRIP_ID_RAW)));
                pidData.setTripId(c.getInt(c.getColumnIndex(TABLES.PID.KEY_TRIP_ID)));
                pidData.setPids(c.getString(c.getColumnIndex(TABLES.PID.KEY_PIDS)));
                pidData.setMileage(c.getDouble(c.getColumnIndex(TABLES.PID.KEY_MILEAGE)));
                pidData.setCalculatedMileage(c.getDouble(c.getColumnIndex(TABLES.PID.KEY_CALCULATED_MILEAGE)));
                pidData.setDeviceId(c.getString(c.getColumnIndex(TABLES.PID.KEY_DEVICE_ID)));

                pidDataEntries.add(pidData);
            } while (c.moveToNext());
        }
        c.close();
        return pidDataEntries;
    }

    /**
     * Number of pid entries
     */
    public int getPidDataEntryCount() {
        String selectQuery = "SELECT * FROM " + TABLES.PID.TABLE_NAME;

        SQLiteDatabase db = databaseHelper.getReadableDatabase();
        Cursor c = db.rawQuery(selectQuery, null);
        int count = c.getCount();
        c.close();
        return count;
    }

    /**
     * Clear all pid data entries
     */
    synchronized public void deleteAllPidDataEntries() {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();

        try {
            db.delete(TABLES.PID.TABLE_NAME, null, null);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    synchronized public void deletePidEntries(List<Pid> pids){
        Log.d(TAG,"deletePidEntries() pids: "+pids);
        SQLiteDatabase db = databaseHelper.getWritableDatabase();

        try {
            for (int i=0;i<pids.size();i++){
                String tripIdRaw = String.valueOf(pids.get(i).getTripIdRaw());
                db.delete(TABLES.PID.TABLE_NAME, TABLES.PID.KEY_TRIP_ID_RAW + "=?"
                        , new String[] { tripIdRaw });
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void deleteAllRows(){
        SQLiteDatabase db = databaseHelper.getWritableDatabase();

        db.delete(TABLES.PID.TABLE_NAME, null, null);
    }

}

File: app/src/pitstop/java/com/pitstop/database/LocalPidStorage.kt
package com.pitstop.database

import android.content.ContentValues
import android.database.Cursor
import android.util.Log
import com.pitstop.bluetooth.dataPackages.PidPackage
import com.pitstop.models.PidGraphDataPoint
import rx.Observable
import java.sql.Date

/**
 * Created by Karol ZDebel on 8/7/2018.
 */
class LocalPidStorage(private val databaseHelper: LocalDatabaseHelper) {

    private val TAG = javaClass.simpleName

    companion object {

        // PID_DATA table create statement
        val CREATE_TABLE_PID_DATA = ("CREATE TABLE IF NOT EXISTS "
                + TABLES.PID.TABLE_NAME + "(" + TABLES.COMMON.KEY_ID + " INTEGER PRIMARY KEY,"
                + TABLES.PID.KEY_TYPE +" TEXT,"
                + TABLES.PID.KEY_VALUE +" TEXT,"
                + TABLES.PID.KEY_RTC_TIME +" INTEGER,"
                + TABLES.COMMON.KEY_CREATED_AT + " TIMESTAMP" + ")")
    }

    fun store(pid: PidPackage): Int {
        val db = databaseHelper.briteDatabase

        val transaction = db.newTransaction()

        var storedCount = 0
        try{
            pid.pids.forEach({
                val values = ContentValues()
                values.put(TABLES.PID.KEY_TYPE, it.key)
                values.put(TABLES.PID.KEY_VALUE, it.value)
                values.put(TABLES.PID.KEY_RTC_TIME, pid.timestamp)
                values.put(TABLES.COMMON.KEY_CREATED_AT, System.currentTimeMillis())
                if (db.insert(TABLES.PID.TABLE_NAME, values) > 0)
                    storedCount = storedCount.inc()
            })
            transaction.markSuccessful()
        }finally{
            transaction.end()
        }

        return storedCount
    }

    fun getAllSync(after: Long): List<PidGraphDataPoint>{
        return cursorToPidGraphPointList(databaseHelper.readableDatabase.query(TABLES.PID.TABLE_NAME
                ,null,TABLES.COMMON.KEY_CREATED_AT+" > ?"
                , arrayOf(after.toString()),null,null,null))
    }

    fun getAll(after: Long): Observable<List<PidGraphDataPoint>>{
        Log.d(TAG,"after: ${Date(after)}")
        return databaseHelper.briteDatabase.createQuery(
                TABLES.PID.TABLE_NAME,"SELECT * FROM "
                +TABLES.PID.TABLE_NAME
                +" WHERE "+TABLES.COMMON.KEY_CREATED_AT+" > " + after
                +" ORDER BY "+TABLES.PID.KEY_RTC_TIME).map {
                    val data = it.run()
                    var pidGraphPointList : List<PidGraphDataPoint> = arrayListOf<PidGraphDataPoint>()
                    Log.d(TAG,"got rows: ${data?.count}")
                    if (data != null) {
                        pidGraphPointList = cursorToPidGraphPointList(data)
                    }
                pidGraphPointList
        }
    }

    fun cursorToPidGraphPointList(cursor: Cursor): List<PidGraphDataPoint>{
        val pidGraphPointList = arrayListOf<PidGraphDataPoint>()
        if (cursor.moveToFirst()){
            while(!cursor.isAfterLast){
                val timestamp = cursor.getLong(cursor.getColumnIndex(TABLES.PID.KEY_RTC_TIME))
                val type = cursor.getString(cursor.getColumnIndex(TABLES.PID.KEY_TYPE))
                val value = cursor.getString(cursor.getColumnIndex(TABLES.PID.KEY_VALUE))
                pidGraphPointList.add(PidGraphDataPoint(timestamp,type,Integer.valueOf(value,16)))
                cursor.moveToNext()
            }
        }
        cursor.close()
        return pidGraphPointList

    }

    fun deleteAllRows() {
        val db = databaseHelper.writableDatabase
        db.delete(TABLES.PID.TABLE_NAME, null, null)
    }
}
